src/actors/symbolication.rs

```diff
@@ -97,37 +98,211 @@ type ComputationChannel = Shared<oneshot::Receiver<(Instant, SymbolicationRespon
+            );
+            Default::default()
+        });
+        cfi_module.scanned = true;
+    }
+
+    /// Processes the [`CfiModule::scanned`] information to update the final module status.
+    ///
+    /// All modules which were not scanned should have [`ObjectFileStatus::Unused`].  This
+    /// processes all the [`CfiModule`]s and updates their [`CfiModule::cfi_status`] as
+    /// required.
+    fn finalize(mut self) -> BTreeMap<CodeModuleId, CfiModule> {
+        for cfi_module in self.inner.values_mut() {
+            if !cfi_module.scanned {
+                cfi_module.cfi_status = ObjectFileStatus::Unused;
+            }
+        }
+        self.inner
+    }
+}
+
+/// A module which was referenced in a minidump and processing information for it.
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct CfiModule {
+    /// Combined features provided by all the DIFs we found for this module.
+    features: ObjectFeatures,
+    /// Status of the CFI or unwind information for this module.
+    cfi_status: ObjectFileStatus,
+    /// Path to the CFI file in our cache, if there was a cache.
+    cfi_path: Option<PathBuf>,
+    /// Indicates if the module was used for stack scanning.
+    scanned: bool,
+}
+
+/// Builder object to collect modules from the minidump.
+///
+/// This collects the modules found by the minidump processor and constructs the
+/// [`CompleteObjectInfo`] objects from them, which are used to eventually return the
+/// module list in the [`SymbolicateStacktraces`] object of the final JSON API response.
+///
+/// The builder requires marking modules that are referenced by stack frames.  This allows it to
+/// omit modules which do not look like real modules (i.e. modules which don't have a valid
+/// code or debug ID, they could be mmap'ed fonts or other such things) as long as they are
+/// not mapped to address ranges used by any frames in the stacktraces.
+struct ModuleListBuilder {
     inner: Vec<(CompleteObjectInfo, bool)>,
 }
 
-impl CodeModulesBuilder {
-    /// Creates a new [`CodeModulesBuilder`] from a list of code modules.
-    pub fn new<I>(iter: I) -> Self
-    where
-        I: IntoIterator<Item = CompleteObjectInfo>,
-    {
-        let mut inner = iter
+impl ModuleListBuilder {
+    fn new(
+        mut cfi_caches: CfiCacheModules,
+        minidump_process_state: &ProcessState,
+        executable_type: ObjectType,
+    ) -> Self {
+        // Firstly mark modules for which we needed CFI information during unwinding.  This
+        // will insert entries into cfi_cache for modules which were missing.
+        for thread in minidump_process_state.threads() {
+            for frame_pair in thread.frames().windows(2) {
+                if let [prev_frame, next_frame] = frame_pair {
+                    if let Some(code_id) = prev_frame.module().and_then(|m| m.id()) {
+                        if matches!(next_frame.trust(), FrameTrust::Scan | FrameTrust::CFIScan) {
```

Comment:
@jan-auer I think your original code forgot to add `FrameTrust::CFI` to this match but this wasn't visible because you only used this information for missing frames instead of for all.  But we want to mark all modules which either used or should have used CFI.  (perhaps the method should be renamed to `mark_cfi_required()` instead of the current `mark_scanned`)

Reply:
https://github.com/getsentry/breakpad/blob/2aaba88045393a024b0e8d285fd2a416c4e05ff4/src/processor/stackwalker_x86.cc#L138 is where these values get assigned btw.

Reply:
No, the intention is to look for the inverse: missing CFI. We detect this by searching for scan trusts.